File: srclab-common/src/main/java/xyz/srclab/common/reflect/MethodKit.java
package xyz.srclab.common.reflect;

import xyz.srclab.annotation.Immutable;
import xyz.srclab.annotation.Nullable;
import xyz.srclab.common.collection.ListKit;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.List;

public class MethodKit {

    @Nullable
    public static Method getMethod(Class<?> cls, String methodName, Class<?>... parameterTypes) {
        try {
            return cls.getMethod(methodName, parameterTypes);
        } catch (NoSuchMethodException e) {
            return null;
        }
    }

    @Nullable
    public static Method getDeclaredMethod(Class<?> cls, String methodName, Class<?>... parameterTypes) {
        try {
            return cls.getDeclaredMethod(methodName, parameterTypes);
        } catch (NoSuchMethodException e) {
            return null;
        }
    }

    @Immutable
    public static List<Method> getMethods(Class<?> cls) {
        return ListKit.immutable(cls.getMethods());
    }

    @Immutable
    public static List<Method> getDeclaredMethods(Class<?> cls) {
        return ListKit.immutable(cls.getDeclaredMethods());
    }

    @Nullable
    public static Object invoke(Method method, @Nullable Object object, Object... args) {
        return invoke(method, object, args, false);
    }

    @Nullable
    public static Object invoke(Method method, @Nullable Object object, Object[] args, boolean force) {
        try {
            if (force) {
                method.setAccessible(true);
            }
            return method.invoke(object, args);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static boolean canOverride(Method method, Class<?> cls) {
        int modifiers = method.getModifiers();
        if (Modifier.isStatic(modifiers)
                || Modifier.isPrivate(modifiers)
                || Modifier.isFinal(modifiers)) {
            return false;
        }
        if (Modifier.isPublic(modifiers)) {
            return true;
        }
        Class<?> declaring = method.getDeclaringClass();
        if (Modifier.isProtected(modifiers)) {
            return declaring.isAssignableFrom(cls);
        }
        return declaring.getPackage().equals(cls.getPackage());
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/reflect/MethodKit.kt
@file:JvmName("MethodKit")
@file:JvmMultifileClass

package xyz.srclab.common.reflect

import java.lang.reflect.Method
import java.lang.reflect.Modifier

@JvmOverloads
fun Class<*>.findMethod(
    name: String,
    declared: Boolean = false,
    deep: Boolean = false,
    vararg parameterTypes: Class<*>
): Method? {
    var method = try {
        this.getMethod(name, *parameterTypes)
    } catch (e: NoSuchFieldException) {
        null
    }
    if (method === null && !declared) {
        return null
    }
    method = this.findDeclaredMethod(name, *parameterTypes)
    if (method === null && !deep) {
        return null
    }
    var superClass = this.superclass
    while (superClass !== null) {
        method = this.findDeclaredMethod(name, *parameterTypes)
        if (method !== null) {
            return method
        }
        superClass = superclass.superclass
    }
    return null
}

fun Class<*>.findDeclaredMethod(methodName: String, vararg parameterTypes: Class<*>): Method? {
    return try {
        this.getDeclaredMethod(methodName, *parameterTypes)
    } catch (e: NoSuchMethodException) {
        null
    }
}

fun Class<*>.findMethods(): List<Method> {
    return this.methods.asList()
}

fun Class<*>.findDeclaredMethods(): List<Method> {
    return this.declaredMethods.asList()
}

@JvmOverloads
fun Method.invoke(owner: Any? = null, force: Boolean = false, vararg args: Any?): Any {
    return try {
        if (force) {
            this.isAccessible = true
        }
        this.invoke(owner, *args)
    } catch (e: Exception) {
        throw IllegalStateException(e)
    }
}

fun Method.isOpen(): Boolean {
    val modifiers = this.modifiers
    if (Modifier.isStatic(modifiers)
        || Modifier.isPrivate(modifiers)
        || Modifier.isFinal(modifiers)
    ) {
        return false
    }
    return Modifier.isPublic(modifiers)
}

fun Method.isOpenFor(cls: Class<*>): Boolean {
    if (!this.isOpen()) {
        return false
    }
    val declaring = this.declaringClass
    return if (Modifier.isProtected(modifiers)) {
        declaring.isAssignableFrom(cls)
    } else declaring.getPackage() == cls.getPackage()
}